### PR TITLE
Add Sparky V1 Variant, modify boards.txt

### PR DESCRIPTION
**Pull request by geosmall**
**Summary**

Pull Request to add Sparky V1 Variant.  First documented here http://www.stm32duino.com/viewtopic.php?f=49&t=4542

This PR fixes/implements the following **bugs/features**

* [ x] Feature 1 - Add Sparky V1 Flight Controller board as new Variant. 

Motivation for making this change:

I would like to propose a new STM core variant based on Sparky V1 Flight Controller. This would enable support for open source drone flight controller development and sharing. It would also enabledevelopment of higher performance open source UAVCAN nodes for drone use.

Reasons for targeting this design:
- Open source design with published schematic
- Based on STM32F303CC processor with FPU
- Available for purchase
- Integrated CAN transceiver

Validation:

See http://www.stm32duino.com/viewtopic.php?p=53147#p53147

### CODE REVIEW

**Comment by fpistm** [comment]
Hi @geosmall 
thanks for this variant.
Please, Could you rebase it on top of the master.
Maybe a new more generic name could be used in the boards.txt: "Flight Controllers" ?

**Comment by geosmall** [comment]
Hi @fpistm 
I've resolved conflict in boards.txt, Travis CI shows a check fail I am not sure how to resolve (I must admit to being a GitHib novice).

There are so many STM32 based flight controllers out there, I targeted this specific board as it is open source and documented so I would propose specific nomenclature to avoid confusion.

**Comment by fpistm** [comment]
Thanks @geosmall 
CI build is failed because it's build under Linux which is case sensitive.
In boards.txt the variant is not correct this should be: `SPARKY_F303CC`

File: variants/SPARKY_F303CC/ldscript.ld [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+**
+**                Set memory bank area and size if external memory is used.
+**
+**  Target      : STMicroelectronics STM32
+**
+**
+**  Distribution: The file is distributed as is, without any warranty
+**                of any kind.
+**
+**  (c)Copyright Ac6.
```
Comment (fpistm) [inline variants/SPARKY_F303CC/ldscript.ld]: Please remove this copyright which is not ok for merge.
You can copy the one from one other variant ldscript.ld.

File: variants/SPARKY_F303CC/stm32f3xx_hal_conf.h [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+/* Exported types ------------------------------------------------------------*/
+/* Exported constants --------------------------------------------------------*/
+
+/* ########################## Module Selection ############################## */
+/**
+  * @brief This is the list of modules to be used in the HAL driver
+  */
+#define HAL_MODULE_ENABLED
+#define HAL_ADC_MODULE_ENABLED
+// #define HAL_CAN_MODULE_ENABLED
```
Comment (fpistm) [inline variants/SPARKY_F303CC/stm32f3xx_hal_conf.h]: If you need CAN, you can enable it by default

File: variants/SPARKY_F303CC/PeripheralPins.c [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+ * Note: Commented lines are alternative possibilities which are not used per default.
+ *       If you change them, you will have to know what you do
+ * =====
+ */
+
+//*** ADC ***
+
+#ifdef HAL_ADC_MODULE_ENABLED
+const PinMap PinMap_ADC[] = {
+  {PA_0,  ADC1,  STM_PIN_DATA_EXT(STM_MODE_ANALOG, GPIO_NOPULL, 0, 1, 0)}, // ADC1_IN1
```
Comment (fpistm) [inline variants/SPARKY_F303CC/PeripheralPins.c]: You can comment lines which is not defined as an Ax in variant.h.

File: boards.txt [hunk 4]
```diff
@@ -897,8 +928,14 @@ GenF4.menu.xserial.none.build.xSerial=-DHAL_UART_MODULE_ENABLED -DHWSERIAL_NONE
 GenF4.menu.xserial.disabled=Disabled (no Serial support)
 GenF4.menu.xserial.disabled.build.xSerial=
 
-RAK.menu.xserial.generic=Enabled (generic 'Serial')
-RAK.menu.xserial.none=Enabled (no generic 'Serial')
+Sparky.menu.xserial.generic=Enabled with generic Serial
```
Comment (fpistm) [inline boards.txt]: Serial menu has been updated.
Could you align with this new definition:
https://github.com/stm32duino/Arduino_Core_STM32/pull/424/commits/8800a2deb89ae6381f700425673bdf590a8900be

Note that you can add USB also if needed

**Comment by geosmall** [comment]
Hi @fpistm ,

I've updated to the requested changes, now appear to be failing an AStyle check, unsure how to best resolve.

Thanks,
George

**Comment by fpistm** [comment]
Thanks @geosmall 
About astyle check, you have to use astyle to format your code.
The astyle config used for the core is available here:
https://github.com/stm32duino/Arduino_Core_STM32/blob/master/CI/astyle/.astylerc

Simply install astyle:
https://sourceforge.net/projects/astyle/files/
then launch it like this:
`astyle --project=.astylerc <your target dir/file>`
or use the python script I provide to ease astyle usage on git root repo:
`python CI/astyle/astyle.py `
note you can use -p option to specify astyle install dir.

File: variants/SPARKY_F303CC/PeripheralPins.c [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+  {NC,    NP,    0}
+};
+#endif
+
+//*** PWM ***
+
+#ifdef HAL_TIM_MODULE_ENABLED
+const PinMap PinMap_PWM[] = {
+  {PA_0,  TIM2,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF1_TIM2, 1, 0)}, // TIM2_CH1
+  {PA_1,  TIM2,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF1_TIM2, 2, 0)}, // TIM2_CH2
```
[the remark this answers is not in the document]
Comment (fpistm) [reply variants/SPARKY_F303CC/PeripheralPins.c]: Still duplicated pins to comment.

**Comment by fpistm** [comment]
If this can help, I've rebased and format your PR:
https://github.com/fpistm/Arduino_Core_STM32/tree/pr-432

**Comment by BennehBoy** [comment]
Haha, I'm amused that Astyle is catching you out also @fpistm

**Comment by fpistm** [comment]
> 
> 
> Haha, I'm amused that Astyle is catching you out also @fpistm

I know this will require some support/question but I think is a necessary evil :smiling_imp:

**Comment by geosmall** [comment]
Let me make sure I'm thinking about this correctly:

On duplicate pins, some of the pins may be used for more than one function across multiple sketches.  For example PA2 might be used as analog input in one sketch (using ADC1_IN3), as PWM input measurement pin in another sketch (using TIM2_CH3), and as hardware USART TX in a third sketch (using USART2_TX).  So it would be expected for PA2 to have multiple uncommented options in PeripheralPins.c.  Does this sound correct?  This is why some pins have duplicate uncommented out lines.

Does this sound correct?

Thanks,
George

**Comment by fpistm** [comment]
Across all array this is not a problem. One pins can have several feature (I2C, UART, SPI, TIM,...)
When I talk about duplicated pins, I talk about the same pin for a dedicated features.
Ex: PA_7 for Timer:
```
  {PA_7,  TIM1,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF6_TIM1, 1, 1)}, // TIM1_CH1N
  {PA_7,  TIM3,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF2_TIM3, 2, 0)}, // TIM3_CH2
  {PA_7,  TIM8,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF4_TIM8, 1, 1)}, // TIM8_CH1N
  {PA_7,  TIM17,  STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF1_TIM17, 1, 0)}, // TIM17_CH1
```
The core will search if PA7 has TIM capabilities then it will get the first one defined. PA7 with TIM1. All other will be ignored but consume memory space for nothing.
By commenting, you are able to choose which TIM use with PA7, ex to use TIM8
```
  //{PA_7,  TIM1,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF6_TIM1, 1, 1)}, // TIM1_CH1N
  //{PA_7,  TIM3,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF2_TIM3, 2, 0)}, // TIM3_CH2
  {PA_7,  TIM8,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF4_TIM8, 1, 1)}, // TIM8_CH1N
  //{PA_7,  TIM17,  STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF1_TIM17, 1, 0)}, // TIM17_CH1
```

**Comment by geosmall** [comment]
Ah, ok thanks will fix that.

**Comment by geosmall** [comment]
I believe I have completed the requested changes and successfully passed CI checks.

File: boards.txt [hunk 6]
```diff
@@ -897,8 +924,14 @@ GenF4.menu.xserial.none.build.xSerial=-DHAL_UART_MODULE_ENABLED -DHWSERIAL_NONE
...
 
-RAK.menu.xserial.generic=Enabled (generic 'Serial')
-RAK.menu.xserial.none=Enabled (no generic 'Serial')
+Sparky.menu.xserial.generic=Enabled (generic 'Serial')
+Sparky.menu.xserial.none=Enabled (no generic 'Serial')
+Sparky.menu.xserial.none.build.xSerial=-DHAL_UART_MODULE_ENABLED -DHWSERIAL_NONE
+Sparky.menu.xserial.disabled=Disabled (no Serial support)
+Sparky.menu.xserial.disabled.build.xSerial=
+
+RAK.menu.xserial.generic=Enabled with generic Serial
```
Comment (fpistm) [inline boards.txt]: Those 2 lines changes must be revert.

Comment (geosmall) [reply boards.txt]: Oops, fixed.

**Comment by geosmall** [comment]
Thanks for merging @fpistm , and thanks for the patience (first GitHub PR :)

**Comment by fpistm** [comment]
Welcome.
All contributions are welcome. This is a community project ;)

**Pull request merged**